ouroboros-network/test/Test/Mux.hs

```diff
@@ -390,61 +517,275 @@ prop_mux_starvation response0 response1 =
+     => Chain block -> [ChainUpdate block] -> DiffTime -> m Property
+demo chain0 updates delay = do
+    client_w <- atomically $ newTBQueue 10
+    client_r <- atomically $ newTBQueue 10
+    let sduLen = 1280
+    let server_w = client_r
+        server_r = client_w
+    producerVar <- atomically $ newTVar (CPS.initChainProducerState chain0)
+    consumerVar <- atomically $ newTVar chain0
+    consumerDone <- atomically newEmptyTMVar
+
+    let Just expectedChain = Chain.applyChainUpdates updates chain0
+        target = Chain.headPoint expectedChain
+        client_mps ChainSync1 = Mx.MiniProtocolDescription
+                                (Just $ consumerInit consumerDone target consumerVar)
+                                Nothing
+        server_mps ChainSync1 = Mx.MiniProtocolDescription
+                                Nothing
+                                (Just $ producerRsp producerVar)
+
+    clientAsync <- startMuxSTM client_mps client_w client_r sduLen Nothing
     serverAsync <- startMuxSTM server_mps server_w server_r sduLen Nothing
 
-    return (server_r, serverAsync)
+    updateAid <- async $ sequence_
+        [ do threadDelay delay -- X milliseconds, just to provide interest
+             atomically $ do
+                 p <- readTVar producerVar
+                 let Just p' = CPS.applyChainUpdate update p
+                 writeTVar producerVar p'
+             | update <- updates
+        ]
+
+    wait updateAid
+    r <- waitBoth clientAsync serverAsync
+    case r of
+         (Just _, _) -> return $ property False
+         (_, Just _) -> return $ property False
+         _           -> do
+             ret <- atomically $ takeTMVar consumerDone
+             return $ property ret
+
+  where
+    checkTip target consumerVar = atomically $ do
+          chain <- readTVar consumerVar
+          return (Chain.headPoint chain == target)
+
+    consumerClient :: Point block -> TVar m (Chain block) -> Client block m ()
+    consumerClient target consChain =
+      Client
+        { rollforward = \_ -> checkTip target consChain >>= \b ->
+            if b then pure $ Left ()
+                 else pure $ Right $ consumerClient target consChain
+        , rollbackward = \_ _ -> checkTip target consChain >>= \b ->
+            if b then pure $ Left ()
+                 else pure $ Right $ consumerClient target consChain
+        , points = \_ -> pure $ consumerClient target consChain
+        }
+
+    consumerInit :: TMVar m Bool -> Point block -> TVar m (Chain block)
+                 -> Channel m BL.ByteString -> m ()
+    consumerInit done_ target consChain channel = do
+       let consumerPeer = chainSyncClientPeer (chainSyncClientExample consChain
+                                               (consumerClient target consChain))
+
+       runPeer nullTracer codecChainSync channel consumerPeer
+       atomically $ putTMVar done_ True
+
+       return ()
+
+    producerRsp ::  TVar m (CPS.ChainProducerState block)
```

Comment:
And I'd call this one `runChainSyncServer`, dual to `runChainSyncClient` - just a suggestion.

Reply:
Will rename.